crates/core/tedge_agent/src/operation_file_cache/mod.rs

```diff
@@ -0,0 +1,349 @@
+use tedge_api::mqtt_topics::ChannelFilter;
+use tedge_api::mqtt_topics::EntityFilter;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::OperationType;
+use tedge_api::path::DataDir;
+use tedge_api::CommandStatus;
+use tedge_downloader_ext::DownloadRequest;
+use tedge_downloader_ext::DownloadResult;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+use tedge_mqtt_ext::TopicFilter;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+type IdDownloadRequest = (String, DownloadRequest);
+type IdDownloadResult = (String, DownloadResult);
+
+fan_in_message_type!(FileCacheInput[MqttMessage, IdDownloadResult]: Debug);
+
+pub struct FileCacheActor {
+    input_receiver: LoggingReceiver<FileCacheInput>,
+    mqtt_sender: DynSender<MqttMessage>,
+    downloader_sender: DynSender<IdDownloadRequest>,
+
+    tedge_http_host: Arc<str>,
+    mqtt_schema: MqttSchema,
+
+    pending_operations: HashMap<String, ConfigUpdateCmdPayload>,
+}
+
+#[async_trait]
+impl Actor for FileCacheActor {
+    fn name(&self) -> &str {
+        "FileCacheActor"
+    }
+
+    async fn run(mut self) -> Result<(), RuntimeError> {
+        while let Some(message) = self.input_receiver.recv().await {
+            match message {
+                FileCacheInput::MqttMessage(message) => {
+                    self.process_mqtt_message(message).await?;
+                }
+                FileCacheInput::IdDownloadResult(message) => {
+                    self.process_download(message).await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl FileCacheActor {
+    async fn process_mqtt_message(
+        &mut self,
+        mqtt_message: MqttMessage,
+    ) -> Result<(), RuntimeError> {
+        let Ok((entity, channel)) = self.mqtt_schema.entity_channel_of(&mqtt_message.topic) else {
+            return Ok(());
+        };
+
+        let Channel::Command {
+            operation: OperationType::ConfigUpdate,
+            cmd_id,
+        } = channel
+        else {
+            return Ok(());
+        };
+
+        let Ok(update_payload) =
+            serde_json::from_slice::<ConfigUpdateCmdPayload>(mqtt_message.payload.as_bytes())
+        else {
+            warn!("Received config update, but payload is malformed");
+            return Ok(());
+        };
+
+        if update_payload.remote_url.is_empty() || update_payload.tedge_url.is_some() {
+            return Ok(());
+        }
+
+        match &update_payload.status {
+            CommandStatus::Executing => {
+                self.download_config_file_to_cache(&mqtt_message.topic, &update_payload)
+                    .await?;
+            }
+            CommandStatus::Successful => {
+                self.delete_symlink_for_config_update(
+                    &entity,
+                    &update_payload.config_type,
+                    &cmd_id,
+                )
+                .unwrap();
```

Comment:
Is it really safe?

Reply:
changed it to return `RuntimeError`, any errors when creating symlinks usually arise from bugs, so they are handled by returning and exiting the actor